apps/omg/test/omg/state/persistence_test.exs

```diff
@@ -33,172 +34,205 @@ defmodule OMG.State.PersistenceTest do
 
     assert {:ok, 1000} ==
-             tx |> Transaction.get_inputs() |> hd() |> InputPointer.Protocol.to_db_key() |> OMG.DB.spent_blknum(db_pid)
+             tx |> Transaction.get_inputs() |> hd() |> InputPointer.Protocol.to_db_key() |> OMG.DB.spent_blknum()
+  end
+
+  @tag fixtures: [:alice]
+  test "exiting utxo is deleted from state", %{alice: alice} do
+    utxo_positions = [
+      Utxo.position(@blknum1, 0, 0),
+      Utxo.position(@blknum1, 0, 1)
+    ]
+
+    [%{owner: alice, currency: @eth, amount: 20, blknum: 1}]
+    |> persist_deposit()
+    |> exec(create_recovered([{1, 0, 0, alice}], @eth, [{alice, 3}]))
+    |> persist_form()
+    |> persist_exit_utxos(utxo_positions)
+
+    :ok = restart_state()
+
+    assert not OMG.State.utxo_exists?(Utxo.position(@blknum1, 0, 0))
+    assert not OMG.State.utxo_exists?(Utxo.position(@blknum1, 0, 1))
   end
 
-  # mimics `&OMG.State.init/1`
-  defp state_from(db_pid) do
-    {:ok, height_query_result} = OMG.DB.get_single_value(:child_top_block_number, db_pid)
-    {:ok, utxos_query_result} = OMG.DB.utxos(db_pid)
+  @tag fixtures: [:alice]
+  test "cannot spend just exited utxo", %{alice: alice} do
+    :ok = persist_deposit([%{owner: alice, currency: @eth, amount: 20, blknum: 1}])
 
-    {:ok, state} = Core.extract_initial_state(utxos_query_result, height_query_result, @interval)
+    {:ok, _, _} = OMG.State.exit_utxos([Utxo.position(1, 0, 0)])
 
-    state
+    # exit db_updates won't get persisted yet, but alice tries to spent it immediately
+    assert :utxo_not_found == exec(create_recovered([{1, 0, 0, alice}], @eth, [{alice, 20}]))
+
+    # retry above with empty in-memory utxoset
+    :ok = restart_state()
+
+    {:ok, _, _} = OMG.State.exit_utxos([Utxo.position(1, 0, 0)])
+    assert :utxo_not_found == exec(create_recovered([{1, 0, 0, alice}], @eth, [{alice, 20}]))
   end
 
-  defp persist_deposit(state, deposits, db_pid) do
-    {:ok, {_, db_updates}, state} = Core.deposit(deposits, state)
-    persist_common(state, db_updates, db_pid)
+  defp persist_deposit(deposits) do
+    {:ok, db_updates} =
+      deposits
+      |> make_deposits()
+      |> OMG.State.deposit()
+
+    :ok = OMG.DB.multi_update(db_updates)
   end
 
-  defp persist_form(state, db_pid) do
-    {:ok, {_, _, db_updates}, state} = Core.form_block(@interval, state)
-    persist_common(state, db_updates, db_pid)
+  defp persist_form(:ok), do: persist_form()
+
+  defp persist_form() do
+    OMG.State.form_block()
+
+    # because `form_block` is non-blocking operation we need to wait it finishes
+    # easiest to do this is to schedule another operation on `OMG.State` which is blocking
+    _ = OMG.State.utxo_exists?(Utxo.position(0, 0, 0))
+
+    :ok
   end
 
-  defp persist_exit_utxos(state, utxo_positions, db_pid) do
-    assert {:ok, {db_updates, _}, state} = utxo_positions |> Core.exit_utxos(state)
-    persist_common(state, db_updates, db_pid)
+  defp exec(:ok, tx), do: exec(tx)
```

Comment:
oh, wait, in the previous `persistence_test` we were piping to transfer `state`, now we don't have anything to pipe through the sequence of steps. So you ended up extending the piped functions to handle the `:ok`'s, but you could just unpipe the whole thing :joy:. I feel a bit culpable for this old test, so if you like, I can take care of it.

oh, so I also had the thought to merge these tests into `state_test.exs` and tidy that up too, so that there is something resembling a proper test of `OMG.State`. But you can punt this on me too, just LMK

Reply:
At first I unpiped the flow, but then I saw the pipe approach as  `assert the previous  operation succeeded` because we don't have the `:error` conterpart